controller/bindings/curl/curl.ml

```diff
@@ -0,0 +1,101 @@
+type error =
+  | UnsuccessfulStatus of int * string
+  | UnreadableStatus of string
+  | ProcessExit of int * string
+  | ProcessKill of int
+  | ProcessStop of int
+
+let pretty_print_error error =
+  match error with
+  | UnsuccessfulStatus (code, body) ->
+      Printf.sprintf "unsuccessful status %d: %s" code body
+
+  | UnreadableStatus body ->
+      Printf.sprintf "unreadable status code %s" body
+
+  | ProcessExit (_, err) ->
+      String.trim err
+
+  | ProcessKill n ->
+      Printf.sprintf "curl killed by signal %d" n
+
+  | ProcessStop n ->
+      Printf.sprintf "curl stopped by signal %d" n
+
+type result =
+  | RequestSuccess of int * string
+  | RequestFailure of error
+
+let exec cmd =
```

Comment:
I think it might be prudent to put this whole function body into a try-with block. According to the [Lwt docs](https://ocsigen.org/lwt/4.5.0/api/Lwt_io):

> input functions of this module raise `End_of_file` when the end-of-file is reached (i.e. when the read function returns 0). Other exceptions are ones caused by the backend read/write functions, such as `Unix.Unix_error`. `exception Channel_closed of string` Exception raised when a channel is closed.

The calls to `Unix.pipe` can themselves raise exceptions.

So it seems we need to expect at least the following exceptions:

- `Unix.Unix_error`
-  `End_of_file`
- `Channel_closed`

A similar situation already exists here: https://github.com/dividat/playos/blob/9e9d243be2f027e58cf8b1a7b5e52150d0e1522e/controller/bindings/util/util.ml#L6-L16

Reply:
Related: https://github.com/dividat/playos/pull/59

Reply:
Thanks for your analysis!

Catching exceptions is done in https://github.com/dividat/playos/pull/54/commits/a5d32d007cd913edc56b7e6602890ceb1045242f.

Errors are integrated into curl error type, no logging is directly applied. I’ll check if callers are taking care of the error case.

I am not completely sure it’s necessary to catch exceptions outside of Lwt, with the same mechanism we introduced in request handlers.

Reply:
Logging request errors reaching the captive portal in https://github.com/dividat/playos/pull/54/commits/6d63f9c3b012fea7146b8c9bb7ee367c910766f0.

Logging is taken care of with requests done by `update`, but not for those done by `zerotier` and `label_printer` I think.

Reply:
Logging `zerotier` errors as well in https://github.com/dividat/playos/pull/54/commits/bddcbb0bc70f0094df6c5ddb4e6fc8fb4176d071.

For `label_printer`, it’s already taken care of by the request handler.